Sources/ProcedureKit/Procedure.swift

```diff
@@ -100,44 +122,69 @@ open class Procedure: Operation, ProcedureProtocol {
         }
     }
 
+    @available(OSX 10.10, iOS 8.0, tvOS 8.0, watchOS 2.0, *)
+    open override var qualityOfService: QualityOfService {
+        get { return super.qualityOfService }
+        set {
+            super.qualityOfService = newValue
+            eventQueue.qualityOfService = newValue.qos
+        }
+    }
+
     internal let identifier = UUID()
 
+    internal class ProcedureQueueContext { }
+    internal let queueAddContext = ProcedureQueueContext()
+
+    deinit {
+        // ensure that the Protected Properies are deinitialized within the lock
+        _stateLock.withCriticalScope {
+            self.protectedProperties = nil
+        }
+    }
+
     // MARK: State
 
-    private var _state = ProcedureKit.State.initialized
-    private let _stateLock = NSRecursiveLock()
+    // the state variable to be used *within* the stateLock
+    private var _state = ProcedureKit.State.initialized {
+        willSet(newState) {
+            _log.verbose(message: "\(_state) -> \(newState)")
+            assert(_state.canTransition(to: newState, whenCancelled: _isCancelled), "Attempting to perform illegal cyclic state transition, \(_state) -> \(newState) for operation: \(identity). Ensure that Procedure instances are added to a ProcedureQueue not an OperationQueue.")
+        }
+    }
+
+    fileprivate let _stateLock = PThreadMutex()
```

Comment:
Would it be possible to expose a `stateLock` to subclasses? Perhaps not this one, but a more public one that developers can use for their own asynchronous `Procedure`'s state? e.g. In my Alamofire-wrapping `NetworkProcedure`, it seems like my `Request` value should be locked, as it's set during execution and has methods called during cancellation. Instead of having to keep my own `stateLock` value, perhaps it would be beneficial to have one already available.